src/hotspot/share/opto/mulnode.cpp

```diff
@@ -281,44 +281,76 @@ Node *MulINode::Ideal(PhaseGVN *phase, bool can_reshape) {
   return res;                   // Return final result
 }
 
-//------------------------------mul_ring---------------------------------------
-// Compute the product type of two integer ranges into this node.
-const Type *MulINode::mul_ring(const Type *t0, const Type *t1) const {
-  const TypeInt *r0 = t0->is_int(); // Handy access
-  const TypeInt *r1 = t1->is_int();
+// Helper classes to perform mul_ring() for MulI/MulLNode.
+template <typename IntegerType>
+struct OverflowType : AllStatic {
+  static_assert(IsSame<IntegerType, TypeInt>::value, "must be TypeInteger");
+  static const Type* overflow() { return TypeInt::INT; }
+};
+
+template <>
+struct OverflowType<TypeLong> : AllStatic {
+  static const Type* overflow() { return TypeLong::LONG; }
+};
+
+template<typename IntegerType>
+class IntegerMulRing {
+  typedef typename Conditional<IsSame<TypeInt, IntegerType>::value, jint, jlong>::type NativeType;
+
+  const IntegerType* _left;
+  const IntegerType* _right;
+  NativeType _min_value = IntegerType::MIN->_lo;
+
+
+  // Check overflow by using arithmetic equality: x = a * b <=> x / a = b. Since a and b are integer numbers, the division
+  // x / a does not underflow/overflow since |x / a| <= |x|. If a * b does underflow/overflow, then we'll get a different
+  // result compared to a * b. Special case MIN_VALUE * -1 whose result is MIN_VALUE.
+  bool does_overflow(const NativeType a, const NativeType b) const {
+    NativeType x = java_multiply(a, b);
```

Comment:
Should we check for 0 values to `return false;` immediately before all calculations?
Also may be path results of `java_multiply()` to does_overflow(). Otherwise you execute it twice.

Reply:
I've updated my patch with your suggestion.